### ISSUE TITLE
Storing build instructions externally risks causing them to be lost to humanity

### ISSUE THREAD

**Issue by ideafarm** [open-watcom/open-watcom-v2#170]
Deleting readme.txt after migrating its content into a GitHub wiki might not be a good idea.  When the content was in a readme.txt file, it was part of the repository content and thus would never be separated from that content.  But now, the readme content is kept in a separate place.  In both of the following scenarios, the readme content would be lost to humanity:

SCENARIO 1:  open-watcom/open-watcom-v2 is forked to a successor repository and then deleted.  In this scenario, the repository's content is preserved in the successor repository, but the wiki is not; anything in the wiki is lost when the original repository is deleted.

SCENARIO 2:  GitHub is obsoleted by X-Hub.  open-watcom/open-watcom-v2 is migrated to X-Hub by downloading the repository content as a zip file.  Since the Wiki is not part of the repository content, it is not preserved within the zip file so is lost.

IMO, GitHub should not be used to store any documentation.  All documentation for OpenWatcom should be stored within the repository content, so that it won't get separated and lost.

**Comment by jmalak** [comment]
One note.
readme.txt file was removed also due to its collision with distribution version of readme.txt file which is now created by documentation process(look into distribution for readme.txt file). 

You must differ between what is distribution of OW and what is development environment.
Distribution version contains only binaries, documentation, samples, etc. Nothing related to OW development environment.
I don't think it is absolutely necessary to hold OW development documentation together with development sources. 
I agree that this documentation is fragmented on many places:
- in sub-projects, various form (wgml, txt, html, doc)
- in documentation project as wgml
- was in readme.txt file
- on WEB sites
- in WIKI
The problem with readme.txt is minimal.
Even if it looks like strange step, I agree with Peter that on Wiki it can be improve by anybody (without OW development repository access) and there can be add many useful things by users if they want. Anyway Wiki on GitHub is standard git repository that it is not problem revert any unwanted change or restore it completely or clone it.
Looks on other similar projects how they manage development documentation. I thing it will be probably also fragmented.
I agree that development documentation could be done more compact, but probably no body want left time to improve it.

**Comment by jmalak** [comment]
I update Readme.md to properly point to Wiki for build instructions.

**Comment by pchapin** [comment]
I agree with @ideafarm that it is important for documentation to be preservable and not tied to any one collaborative system such as GitHub. In fact, I'd hold up the wiki on www.openwatcom.org as an example of a great library of useful information that is hard to access and manipulate because it is tied to MediaWiki.

However, the GitHub wikis are themselves Git repositories that can be cloned and manipulated as an ordinary collection of files. It is a simple matter to take the wiki with us should we ever decide to abandon GitHub. It would even be easy to combine the information in the wiki back into the main repository if necessary: just copy the files from one working copy to another and commit.

The advantage of having the README in the wiki, as @jmalak notices, is that anyone can edit it. This invites assistance from people who otherwise don't want to bother with the complications of pull requests, etc. In fact, the wiki can be edited on the GitHub site without even bothering with Git.

Anyway, rest assured, @ideafarm, I totally agree with your sentiment. I wouldn't have moved the documentation to the wiki if I thought doing so would somehow tie the content to GitHub.

**Comment by ideafarm** [comment]
This issue came to light for me because I forked and your wiki didn't come with it.  I'd feel more comfortable if (my wish list):

(1) All variants of OW used the same infrastructure, i.e. GitHub.

(2) All build documentation, including instructions/recommendations for maintaining a fork in a GitHub repository, is kept within the repository content.

(3) The instructions include guidance on how to periodically take a snapshot of the wiki and any other external documentation and to keep these snapshots within the repo content.

(4) Guidance regarding a standard defining a "conformant fork", where the standard is a minimal set of rules that will ensure that the repos of all "conformant" forks are complete, i.e. can be used even if everything external goes up in flames.

**Comment by pchapin** [comment]
I'm not sure what you mean by "my wiki." If you are talking about the repository used to hold the Open Watcom wiki here on GitHub, it's at a different URL. If you go to the [wiki page](https://github.com/open-watcom/open-watcom-v2/wiki) and look to the right you should see a URL under "clone this wiki locally." That's the one to use. Does this address your concerns about storing the information in repository content?

**Comment by ideafarm** [comment]
The fact that snapshots of the external documentation COULD be preserved within the repo does not address my concern.  I want you to make such a snapshot now, so that the repo again contains ALL information needed for its use.  I also want you to periodically make such a snapshot.

Moving documentation content out of the repo is a very bad idea.  All documentation should live within the repo so that it won't get separated and lost.  My concern is that someone will download a zip of the repo thinking that the zip contains everything.  Because you moved some of the doc to live someplace other than within the repo, there is now nothing to ensure that the repo continues to be complete.  This is such a very bad idea that it alone destroys any confidence that I used to have that you and Jiri know what the hell you are doing!  :)

IOW, the fact that the wiki can be cloned does not mean that it WILL be cloned and that it WILL always travel with the repo content.  The only way to eliminate my concern is to put the doc back into the repo.

**Comment by pchapin** [comment]
Since the wiki can be edited by anyone it gives the community a way to contribute to the project without the overhead of dealing with all the development machinery (cloning the source code, etc). Many projects find this beneficial. The wiki on www.openwatcom.org is loaded with good information added by third parties.

The issues list contains valuable information about the project as well but it is also not in the repository. Personally I find that troubling, but I live with it. At least the wiki is easily extracted from GitHub. I suspect the issues list is more difficult to move around.

In other words: it's unrealistic to hope that the repository will be the one and only location of information about the project. You should probably let go of that goal. That said, I do agree that it is important to keep the information as accessible as possible. Personally, I'm a lot more worried about the issues list than the wiki.

**Comment by groessler** [comment]
I agree with @ideafarm, that at least build instructions should be part of the source repository. Every project I know comes with at least one of "README", "INSTALL", "BUILD" files, describing how to build and install the project. It's quite inconvenient to be redirected to a web site while working on the command line. :-)

**Comment by ideafarm** [comment]
Proposal:

(1)  Put a snapshot of the wiki into the repo, with a plain text file stating when the snapshot was taken, stating that it is probably stale, and providing the URL of the source of the snapshot.  Also include a suggestion that a fresh snapshot be taken occasionally.

(2)  Do it for the issues list info, too.  Just use the Chrome browser to download the issues list as a PDF file, if GitHub does not provide a way to export the discussions.

**Comment by jmalak** [comment]
I think note of @groessler has value. We should add some text file, by example build.txt which contains minimal info about how to build OW distribution. But adding of Wiki and other thing into main repository is unrealistic and have no much sense. If and only if we want to move OW v2 to other site then conversion of parts outside development sources to different format will be necessary anyway. I don't want to lost my time to do some standardization of things which are very subjective and each human have different opinion how it should be done. I don't thing that next real fork will be done that thinking about it is little useless. You can do your own fork and try to do what you prefer, it is freedom and advance of Git repositories. 

We are concentrated to do real progress on OW V2 this is difference to official OW site, where was many similar discussion about plans but no real work. If you want to help this project with testing you are welcome. Don't assume enthusiasm for large reworking of OW, similar changes should have very good reason and some gain. But what you now propose don't bother anyone too much.
I am trying to do things simple not complicated and I always think how to do contributors live simple.

I don't plan to move this OW V2 fork to any other site until someone convince me that somewhere is much better environment for OW V2.

**Comment by pchapin** [comment]
I agree with @jmalak  that one of the problems with the original fork is that it is all talk and no action. We need to be careful to avoid that trap.

As far as the location of the build instructions go, it doesn't matter that much to me. On the wiki there is the advantage that people who are not otherwise interested in contributing can still modify them. In the past we've had people who were just building the system for their own use and didn't want to become contributors say, "the build instructions are unclear and should be updated." Now those people can tweak the instructions. How many people are really in that situation? Probably not many.

As far as access from the command line goes. I guess I find the issue hard to understand. Anyone who can clone the main repository can clone the wiki. I have the main repository, the wiki, and the [GitHub Open Watcom web site](http://open-watcom.github.io/open-watcom/) all cloned side by side on my machine. They are all just files in my file system to me. What's the problem?

**Comment by jmalak** [comment]
build.txt we should add only for consistency, but for me is preferred way to put this information into Wiki to simply enhance it by comments and also samples could be put there etc. I agree with @pchapin that we have all info in repositories and therefore we have no reason to complicate thing.

I have setup tracking for all my forked repositories that I simply synchronize them with master repositories regularly by cron script. I have no problem with anything out-dated.

If anybody want to work with full OW V2 fork build then minimal knowledge is expected, for these peoples build.txt could be enough, but this info is not for beginners. For them is much suitable something like Wiki with detail comment, examples, enhanced by highlighting, cross-links, tables etc.

**Comment by ArmstrongJ** [comment]
I actually don't have a problem keeping build instructions in the wiki.  Although many (or most) open-source projects have an *INSTALL* file, this project is significantly larger and more complex.  We're not talking about some silly library or dynamic language interpreter.  People need to remember that Open Watcom is enormous and complex.  The build instructions aren't "./configure && make all" like most projects.  This project is drastically more complicated.  We shouldn't imply that building Open Watcom is easy (you usually need a DOS emulator and Windows HTML Help Workshop to do a full build, after all).

Many, many large projects will have an *INSTALL* file that simply either: 1. redirects users to a website or 2. redirects people to a directory.  GCC, for example, which is a similarly sized project in terms of complexity, has an entire directory describing how to build it.  I think our wiki-based solution is fine.  

That said, if you want to package build instructions with a *source* distribution, i.e. an archive with the entire source tree, then we can think about maybe pulling in the wiki's build instructions.  That can be done using Git "submodules," where a directory would point at another git repo (the wiki, in this case), or the source distro build scripts could pull in the latest wiki build instructions with a simple git command.  Having the build instructions in both the source repo and the wiki will just lead to them being out of sync and one being completely wrong eventually.

Like Jiri mentioned here or another bug, let's try and keep things simple.

**Comment by ideafarm** [comment]
I can see why Jiri didn't get on well with the plodding corporate types.  I mostly agree with Jiri and am 90% like him and 10% like the plodding corp types.  I'm moving as quickly as I can toward getting my new computer set up and then setting up a "QA/QC department" for OW.  But I was arrested and jailed today.  What a PITA.  These fucking police are totally out of control here in Googleville (Mountain View, CA).
